File: src/next.ts
/**
 * "다음에 뭘 할까" 후보 랭킹 — 순수 함수.
 *
 * 보드를 그대로 들이밀면 고를 수 없다(실사용 보드가 이미 30건대다). 게다가 이걸 쓰는
 * 슬래시 커맨드의 선택지 UI 는 한 번에 넉넉히 4개라, **목록을 좁히는 판정**이 먼저 있어야
 * 한다. 그 판정을 출력 포맷·CLI 에서 떼어 둔 이유는 조합이 많아서다 — 상태 × 마감 ×
 * 우선순위 × 계층을 주입 없이 테스트하려면 순수해야 한다(`./doing` 과 같은 이유).
 *
 * 여기서 하지 않는 것: 세션 대조. `doingState` 는 서버가 이미 `GET /api/todos` 응답에
 * 얹어 준다(`./doing` 의 `resolveDoingState`). 이 모듈은 그 판정을 **소비**만 한다.
 */

import type { TodoView } from './refs';

/** 후보 하나 — 랭킹 점수와 사람이 읽는 근거를 얹는다. */
export interface NextCandidate {
  todo: TodoView;
  score: number;
  /** 왜 위로 왔는지 (예: `이어받기(멈춤) · p2`). 선택지 설명에 그대로 쓴다. */
  reason: string;
}

export interface RankNextOptions {
  /** 기준 시각(ms) — 마감 D-day 계산의 "오늘". 테스트가 고정한다. */
  now: number;
  /** 상위 몇 개까지 남길지. 생략하면 전부. */
  limit?: number;
}

/**
 * CLI 가 기본으로 보여줄 후보 수.
 *
 * 선택지 UI 가 4개까지라 고르는 것은 상위 4개지만, 그 위에 무엇이 밀려났는지 보이지 않으면
 * "왜 이 4개인가" 를 사람이 검증할 수 없다. 그래서 목록은 두 배까지 보여준다.
 */
export const NEXT_DEFAULT_LIMIT = 8;

const PRIORITY_SCORE: Record<TodoView['priority'], number> = { p1: 30, p2: 18, p3: 6, p4: 0 };
const PRIORITY_ORDER: Record<TodoView['priority'], number> = { p1: 0, p2: 1, p3: 2, p4: 3 };

/** 마감이 코앞이라고 볼 기간(일). 이 안쪽만 점수를 받는다. */
const DUE_SOON_DAYS = 7;
/** 댓글이 "방금 오갔다" 고 볼 기간(일). 사용자가 보드에서 답을 달았을 가능성이 높은 구간. */
const FRESH_COMMENT_DAYS = 3;

const DAY_MS = 86_400_000;

/** 열려 있는가 — done/보관은 후보가 아니다. */
function isOpen(todo: TodoView): boolean {
  return todo.status !== 'done' && !todo.archivedAt;
}

/** `YYYY-MM-DD` → 에포크 기준 일수. 자리 파싱이라 파싱 타임존에 흔들리지 않는다. */
function dayNumber(date: string): number {
  const [y, m, d] = date.slice(0, 10).split('-').map(Number);
  if (!y || !m || !d) {
    return Number.NaN;
  }
  return Math.floor(Date.UTC(y, m - 1, d) / DAY_MS);
}

/** 기준 시각의 **로컬** 날짜를 같은 일수 축으로 옮긴다 — 마감은 사람이 사는 날짜다. */
function todayNumber(now: number): number {
  const t = new Date(now);
  return Math.floor(Date.UTC(t.getFullYear(), t.getMonth(), t.getDate()) / DAY_MS);
}

interface Scored {
  score: number;
  labels: string[];
}

/**
 * 마감 점수 — 지난 것 > 오늘 > 코앞. 마감이 없거나 깨진 값이면 0점이고 라벨도 없다.
 *
 * 깨진 `due` 를 0점으로 흘리는 것은 의도다. 여기서 던지면 항목 하나의 오타가 "다음 작업"
 * 전체를 못 쓰게 만든다 — 랭킹은 조언이고, 조언은 최악의 입력에서도 나와야 한다.
 */
function dueScore(todo: TodoView, now: number): Scored {
  if (!todo.due) {
    return { score: 0, labels: [] };
  }
  const days = dayNumber(todo.due) - todayNumber(now);
  if (Number.isNaN(days)) {
    return { score: 0, labels: [] };
  }
  if (days < 0) {
    return { score: 70, labels: [`마감 D+${-days}`] };
  }
  if (days === 0) {
    return { score: 60, labels: ['마감 D-day'] };
  }
  return days <= DUE_SOON_DAYS
    ? { score: 35, labels: [`마감 D-${days}`] }
    : { score: 0, labels: [] };
}

/**
 * 진행중 점수 — **주인 없는 doing 이 가장 위**다.
 *
 * 이 순서가 랭킹의 핵심이다. 새 일을 하나 더 벌이는 것보다 "누군가 시작해 놓고 끊긴 것"
 * 을 잇는 편이 거의 항상 낫다. `live`(세션이 지금 그걸 붙들고 있음)는 후보에서 아예
 * 빠지므로 여기 오지 않는다 — 두 에이전트가 한 항목을 같이 잡는 사고를 막는 자리다.
 */
function doingScore(todo: TodoView): Scored {
  if (todo.status !== 'doing') {
    return { score: 0, labels: [] };
  }
  if (todo.doingState === 'gone') {
    return { score: 100, labels: ['이어받기(세션 없음)'] };
  }
  if (todo.doingState === 'idle') {
    return { score: 90, labels: ['이어받기(멈춤)'] };
  }
  // 판정 불가(`unknown`) 나 사람이 잡은 doing. 죽었다고 단정할 수 없으니 "이어받기" 라고
  // 부르지 않되, todo 보다는 위에 둔다 — 이미 손댄 일이다.
  return { score: 45, labels: [todo.doingBy ? `진행중(${todo.doingBy})` : '진행중'] };
}

/** 우선순위 점수. 라벨은 p1/p2 만 — p3/p4 까지 찍으면 근거 줄이 소음이 된다. */
function priorityScore(todo: TodoView): Scored {
  const score = PRIORITY_SCORE[todo.priority];
  return { score, labels: todo.priority === 'p1' || todo.priority === 'p2' ? [todo.priority] : [] };
}

/** 최근 댓글 점수 — 보드에서 대화가 오가는 중이면 그쪽이 대개 지금 관심사다. */
function commentScore(todo: TodoView, now: number): Scored {
  if (!todo.lastCommentAt) {
    return { score: 0, labels: [] };
  }
  const at = Date.parse(todo.lastCommentAt);
  if (Number.isNaN(at) || now - at > FRESH_COMMENT_DAYS * DAY_MS) {
    return { score: 0, labels: [] };
  }
  return { score: 12, labels: ['최근 댓글'] };
}

function scoreOf(todo: TodoView, now: number): Scored {
  const parts = [
    doingScore(todo),
    dueScore(todo, now),
    priorityScore(todo),
    commentScore(todo, now),
  ];
  return {
    score: parts.reduce((sum, part) => sum + part.score, 0),
    labels: parts.flatMap((part) => part.labels),
  };
}

/**
 * 점수 내림차순. 동점은 우선순위 → position → number → ref 로 완전히 결정된다 —
 * 같은 보드를 두 번 물었을 때 순서가 흔들리면 사용자가 1번을 신뢰할 수 없다.
 */
function compareCandidates(a: NextCandidate, b: NextCandidate): number {
  return (
    b.score - a.score ||
    PRIORITY_ORDER[a.todo.priority] - PRIORITY_ORDER[b.todo.priority] ||
    a.todo.position - b.todo.position ||
    a.todo.number - b.todo.number ||
    a.todo.ref.localeCompare(b.todo.ref)
  );
}

/**
 * 착수 후보를 랭킹해 상위 `limit` 개를 돌려준다.
 *
 * 후보에서 빠지는 것 셋:
 * 1. done / 보관됨 — 할 일이 아니다.
 * 2. `doingState: 'live'` — 살아 있는 세션이 지금 붙들고 있다.
 * 3. **열린 자식을 가진 부모** — 우산 항목이다. 실제 착수 대상은 그 자식이라, 부모를
 *    골라도 무엇을 할지가 안 나온다. 자식이 전부 done 인 부모는 남는다(마무리가 남았다).
 */
export function rankNext(todos: readonly TodoView[], options: RankNextOptions): NextCandidate[] {
  const umbrellas = new Set(
    todos.filter((todo) => todo.parentId && isOpen(todo)).map((todo) => todo.parentId as string),
  );
  const candidates = todos
    .filter((todo) => isOpen(todo) && !umbrellas.has(todo.id) && todo.doingState !== 'live')
    .map((todo) => {
      const { score, labels } = scoreOf(todo, options.now);
      return { todo, score, reason: labels.length > 0 ? labels.join(' · ') : '대기 중' };
    })
    .sort(compareCandidates);
  return options.limit === undefined ? candidates : candidates.slice(0, options.limit);
}

/**
 * 후보 목록을 컴팩트하게 렌더한다 — 한 줄에 `번호. ref  제목  — 근거`.
 *
 * 맨숫자가 아니라 완전 참조(`rocky-12`)를 찍는다. 이 목록은 보드를 넘나들 수 있고(`--all`),
 * 사용자가 다음에 하는 일이 그 참조를 그대로 다른 명령에 붙여넣는 것이다.
 */
export function formatNextCandidates(candidates: readonly NextCandidate[]): string {
  if (candidates.length === 0) {
    return '착수할 후보가 없다 — 열린 항목이 없거나, 남은 것이 전부 다른 세션에 잡혀 있다';
  }
  return candidates
    .map((c, i) => `${i + 1}. ${c.todo.ref}  ${c.todo.title}  — ${c.reason}`)
    .join('\n');
}

File: crates/rocky-todo-core/src/next.rs
//! "다음에 뭘 할까" 후보 랭킹 — 순수 함수. TS 원본 `src/next.ts`.
//!
//! 세션 대조는 하지 않는다 — `doing_state` 는 서버가 이미 목록 응답에 얹어 준다.
//! 이 모듈은 그 판정을 **소비**만 한다.

use serde::Serialize;

use crate::doing::DoingState;
use crate::refs::TodoView;
use crate::types::{TodoPriority, TodoStatus};

/// 후보 하나 — 랭킹 점수와 사람이 읽는 근거를 얹는다.
#[derive(Debug, Clone)]
pub struct NextCandidate {
    pub todo: TodoView,
    pub score: i64,
    /// 왜 위로 왔는지 (예: `이어받기(멈춤) · p2`).
    pub reason: String,
}

pub struct RankNextOptions {
    /// 기준 시각(ms) — 마감 D-day 계산의 "오늘". 테스트가 고정한다.
    pub now: i64,
    /// 상위 몇 개까지. None 이면 전부.
    pub limit: Option<usize>,
}

/// CLI 가 기본으로 보여줄 후보 수 — 한 화면에서 훑고 고를 수 있는 상한.
pub const NEXT_DEFAULT_LIMIT: usize = 8;

/// 랭킹 밴드 — **앞 자리가 뒤 자리를 항상 이긴다.** 각 범주를 자리값이 다른 칸(0..99)에
/// 나눠 담아, 하위 범주를 전부 합쳐도 상위 범주의 한 칸을 넘지 못한다.
mod band {
    /// 주인 없는 진행중 — 이어받을 것이 있으면 그게 최우선이다.
    pub const ORPHAN: i64 = 100_000_000;
    pub const DUE: i64 = 1_000_000;
    /// 판정할 수 없는 진행중 — 마감 아래, 우선순위 위.
    pub const DOING: i64 = 10_000;
    pub const PRIORITY: i64 = 100;
    pub const COMMENT: i64 = 1;
}

fn priority_score(p: TodoPriority) -> i64 {
    match p {
        TodoPriority::P1 => 4,
        TodoPriority::P2 => 3,
        TodoPriority::P3 => 2,
        TodoPriority::P4 => 1,
    }
}

fn priority_order(p: TodoPriority) -> i64 {
    match p {
        TodoPriority::P1 => 0,
        TodoPriority::P2 => 1,
        TodoPriority::P3 => 2,
        TodoPriority::P4 => 3,
    }
}

/// 마감이 코앞이라고 볼 기간(일).
const DUE_SOON_DAYS: i64 = 7;
/// 댓글이 "방금 오갔다" 고 볼 기간(일).
const FRESH_COMMENT_DAYS: i64 = 3;

const DAY_MS: i64 = 86_400_000;

/// 열려 있는가 — done/보관은 후보가 아니다.
fn is_open(todo: &TodoView) -> bool {
    todo.todo.status != TodoStatus::Done && todo.todo.archived_at.is_none()
}

/// `YYYY-MM-DD` → 에포크 기준 일수. **달력에 없는 날짜는 None** — `2026-02-31` 같은
/// 값이 있지도 않은 마감으로 D-day 를 찍지 않게 되돌려 대조한다.
fn day_number(date: &str) -> Option<i64> {
    let text: String = date.chars().take(10).collect();
    let bytes = text.as_bytes();
    if bytes.len() != 10 || bytes[4] != b'-' || bytes[7] != b'-' {
        return None;
    }
    let digits_ok = |range: std::ops::Range<usize>| bytes[range].iter().all(u8::is_ascii_digit);
    if !digits_ok(0..4) || !digits_ok(5..7) || !digits_ok(8..10) {
        return None;
    }
    let y: i32 = text[0..4].parse().ok()?;
    let m: u32 = text[5..7].parse().ok()?;
    let d: u32 = text[8..10].parse().ok()?;
    let date = chrono::NaiveDate::from_ymd_opt(y, m, d)?; // 달력 검증까지 한 번에
    Some(date.and_hms_opt(0, 0, 0)?.and_utc().timestamp_millis() / DAY_MS)
}

/// 기준 시각의 **로컬** 날짜를 같은 일수 축으로 — 마감은 사람이 사는 날짜다.
fn today_number(now: i64) -> i64 {
    use chrono::{Datelike, TimeZone};
    let local = chrono::Local
        .timestamp_millis_opt(now)
        .single()
        .unwrap_or_else(chrono::Local::now);
    let (y, m, d) = (local.year(), local.month(), local.day());
    chrono::NaiveDate::from_ymd_opt(y, m, d)
        .and_then(|date| date.and_hms_opt(0, 0, 0))
        .map(|dt| dt.and_utc().timestamp_millis() / DAY_MS)
        .unwrap_or(0)
}

struct Scored {
    score: i64,
    labels: Vec<String>,
}

fn none_scored() -> Scored {
    Scored {
        score: 0,
        labels: Vec::new(),
    }
}

/// 마감 점수 — 지난 것 > 오늘 > 코앞. 깨진 값은 0점으로 흘린다(랭킹은 조언이고,
/// 조언은 최악의 입력에서도 나와야 한다).
fn due_score(todo: &TodoView, now: i64) -> Scored {
    let Some(due) = &todo.todo.due else {
        return none_scored();
    };
    let Some(day) = day_number(due) else {
        return none_scored();
    };
    let days = day - today_number(now);
    if days < 0 {
        Scored {
            score: 3 * band::DUE,
            labels: vec![format!("마감 D+{}", -days)],
        }
    } else if days == 0 {
        Scored {
            score: 2 * band::DUE,
            labels: vec!["마감 D-day".to_string()],
        }
    } else if days <= DUE_SOON_DAYS {
        Scored {
            score: band::DUE,
            labels: vec![format!("마감 D-{days}")],
        }
    } else {
        none_scored()
    }
}

/// 진행중 점수 — **주인 없는 doing 이 가장 위**. `live` 는 후보에서 아예 빠진다.
fn doing_score(todo: &TodoView) -> Scored {
    if todo.todo.status != TodoStatus::Doing {
        return none_scored();
    }
    match todo.doing_state {
        Some(DoingState::Gone) => Scored {
            score: 2 * band::ORPHAN,
            labels: vec!["이어받기(세션 없음)".to_string()],
        },
        Some(DoingState::Idle) => Scored {
            score: band::ORPHAN,
            labels: vec!["이어받기(멈춤)".to_string()],
        },
        // 판정 불가나 사람이 잡은 doing — "이어받기" 라 부르지 않되 todo 보다는 위.
        _ => Scored {
            score: band::DOING,
            labels: vec![match &todo.todo.doing_by {
                Some(by) => format!("진행중({by})"),
                None => "진행중".to_string(),
            }],
        },
    }
}

/// 우선순위 점수. 라벨은 p1/p2 만 — p3/p4 까지 찍으면 근거 줄이 소음이 된다.
fn priority_scored(todo: &TodoView) -> Scored {
    let p = todo.todo.priority;
    let labels = match p {
        TodoPriority::P1 | TodoPriority::P2 => vec![p.as_str().to_string()],
        _ => Vec::new(),
    };
    Scored {
        score: priority_score(p) * band::PRIORITY,
        labels,
    }
}

/// 최근 댓글 점수 — 보드에서 대화가 오가는 중이면 그쪽이 대개 지금 관심사다.
fn comment_score(todo: &TodoView, now: i64) -> Scored {
    let Some(last) = &todo.last_comment_at else {
        return none_scored();
    };
    let Ok(at) = chrono::DateTime::parse_from_rfc3339(last) else {
        return none_scored();
    };
    let at_ms = at.timestamp_millis();
    if now - at_ms > FRESH_COMMENT_DAYS * DAY_MS {
        return none_scored();
    }
    Scored {
        score: band::COMMENT,
        labels: vec!["최근 댓글".to_string()],
    }
}

fn score_of(todo: &TodoView, now: i64) -> Scored {
    let parts = [
        doing_score(todo),
        due_score(todo, now),
        priority_scored(todo),
        comment_score(todo, now),
    ];
    let mut score = 0;
    let mut labels = Vec::new();
    for part in parts {
        score += part.score;
        labels.extend(part.labels);
    }
    Scored { score, labels }
}

/// 점수 내림차순, 동점은 우선순위 → position → number → ref 로 완전히 결정.
fn compare(a: &NextCandidate, b: &NextCandidate) -> std::cmp::Ordering {
    b.score
        .cmp(&a.score)
        .then(priority_order(a.todo.todo.priority).cmp(&priority_order(b.todo.todo.priority)))
        .then(a.todo.todo.position.cmp(&b.todo.todo.position))
        .then(a.todo.todo.number.cmp(&b.todo.todo.number))
        .then(a.todo.r#ref.cmp(&b.todo.r#ref))
}

/// 착수 후보를 랭킹해 상위 limit 개. 빠지는 것: done/보관, `Live` doing,
/// **열린 자식을 가진 부모**(우산 항목 — 자식이 전부 done 인 부모는 남는다).
pub fn rank_next(todos: &[TodoView], options: &RankNextOptions) -> Vec<NextCandidate> {
    let umbrellas: std::collections::HashSet<&str> = todos
        .iter()
        .filter(|t| t.todo.parent_id.is_some() && is_open(t))
        .filter_map(|t| t.todo.parent_id.as_deref())
        .collect();
    let mut candidates: Vec<NextCandidate> = todos
        .iter()
        .filter(|t| {
            is_open(t)
                && !umbrellas.contains(t.todo.id.as_str())
                && t.doing_state != Some(DoingState::Live)
        })
        .map(|t| {
            let Scored { score, labels } = score_of(t, options.now);
            NextCandidate {
                todo: t.clone(),
                score,
                reason: if labels.is_empty() {
                    "대기 중".to_string()
                } else {
                    labels.join(" · ")
                },
            }
        })
        .collect();
    candidates.sort_by(compare);
    match options.limit {
        None => candidates,
        Some(limit) => {
            candidates.truncate(limit);
            candidates
        }
    }
}

/// `summary` 로 자를 길이.
const SUMMARY_MAX: usize = 160;

/// `--json` 이 내보내는 후보 하나 — **TodoView 전체가 아니다** (고르는 데 필요한 것만).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NextCandidateJson {
    pub r#ref: String,
    pub number: i64,
    pub board: String,
    pub title: String,
    /// `NextCandidate.reason` 그대로.
    pub reason: String,
    pub priority: TodoPriority,
    pub status: TodoStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due: Option<String>,
    pub labels: Vec<String>,
    pub comment_count: i64,
    /// description 을 한 줄로 눌러 SUMMARY_MAX 자까지. 없으면 필드 자체가 없다.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
}

/// 공백류를 단일 공백으로 눌러 한 줄로 — 개행 든 제목이 가짜 후보로 보이는 오선택 방지.
fn flatten(text: &str) -> String {
    text.split_whitespace().collect::<Vec<_>>().join(" ")
}

/// 여러 줄 markdown 을 한 줄로 눌러 max 자까지. 자르면 `…`.
fn condense(text: &str, max: usize) -> Option<String> {
    let flat = flatten(text);
    if flat.is_empty() {
        return None;
    }
    let chars: Vec<char> = flat.chars().collect();
    if chars.len() <= max {
        Some(flat)
    } else {
        let cut: String = chars[..max].iter().collect();
        Some(format!("{}…", cut.trim_end()))
    }
}

/// 후보를 `--json` 출력용 컴팩트 형태로. 보드 못 찾으면 빈 문자열(던지지 않는다).
pub fn to_json_candidates(
    candidates: &[NextCandidate],
    board_key_of: impl Fn(&str) -> Option<String>,
) -> Vec<NextCandidateJson> {
    candidates
        .iter()
        .map(|c| NextCandidateJson {
            r#ref: c.todo.r#ref.clone(),
            number: c.todo.todo.number,
            board: board_key_of(&c.todo.todo.board_id).unwrap_or_default(),
            title: c.todo.todo.title.clone(),
            reason: c.reason.clone(),
            priority: c.todo.todo.priority,
            status: c.todo.todo.status,
            due: c.todo.todo.due.clone(),
            labels: c.todo.todo.labels.clone(),
            comment_count: c.todo.comment_count,
            summary: condense(&c.todo.todo.description, SUMMARY_MAX),
        })
        .collect()
}

/// 후보 목록 컴팩트 렌더 — 한 줄에 `번호. ref  제목  — 근거`. **한 후보는 반드시 한 줄.**
pub fn format_next_candidates(candidates: &[NextCandidate]) -> String {
    if candidates.is_empty() {
        return "착수할 후보가 없다 — 열린 항목이 없거나, 남은 것이 전부 다른 세션에 잡혀 있다"
            .to_string();
    }
    candidates
        .iter()
        .enumerate()
        .map(|(i, c)| {
            format!(
                "{}. {}  {}  — {}",
                i + 1,
                c.todo.r#ref,
                flatten(&c.todo.todo.title),
                c.reason
            )
        })
        .collect::<Vec<_>>()
        .join("\n")
}
